### PR TITLE
Missing Functions on edges/nested edges

### PR DESCRIPTION
**Pull request by rachchan**
### What changes were proposed in this pull request?
add the missing functions that were in Edge but not Edges or NestedEdges

### Why are the changes needed?
Missing functionality

### Does this PR introduce any user-facing change? If yes is this documented?
Yes, yes

### How was this patch tested?
Python tests

### Issues

_If this resolves any issues, please link to them here, the format is a KEYWORD followed by @<ISSUE NUMBER>__
KEYWORDS available are `close`, `closes`, `closed`, `fix`, `fixes`, `fixed`, `resolve`, `resolves`, `resolved`.
Please delete this text before creating your PR

### CODE REVIEW

File: raphtory/src/db/graph/edge.rs [hunk 1]
```diff
@@ -51,6 +53,10 @@ impl<G: GraphViewOps> EdgeView<G> {
     pub fn new(graph: G, edge: EdgeRef) -> Self {
         Self { graph, edge }
     }
+
+    pub fn iter(&self) -> Box<dyn Iterator<Item = EdgeView<G>> + Send> {
+        Box::new(iter::once(self.clone()))
+    }
```
Comment (ljeub-pometry) [inline raphtory/src/db/graph/edge.rs:59]: why do we need this?

File: raphtory/src/db/graph/edge.rs [hunk 2]
```diff
@@ -360,6 +417,8 @@ impl<G: GraphViewOps> EdgeListOps for BoxedIter<BoxedIter<EdgeView<G>>> {
     type VList = Box<dyn Iterator<Item = Box<dyn Iterator<Item = VertexView<G>> + Send>> + Send>;
     type IterType<T> = Box<dyn Iterator<Item = Box<dyn Iterator<Item = T> + Send>> + Send>;
 
+    // type WindowedViewType = EdgeView<G>;
+
```
Comment (ljeub-pometry) [inline raphtory/src/db/graph/edge.rs:421]: delete this comment

File: raphtory/src/python/graph/edge.rs [hunk 3]
```diff
@@ -646,13 +892,55 @@ impl PyNestedEdges {
...
+        let edges = self.builder.clone();
+        edges()
+            .layer_names()
+            .map(
+                |e: Box<dyn Iterator<Item = Box<dyn Iterator<Item = ArcStr> + Send>> + Send>| {
+                    e.map(|e| e.collect_vec()).collect_vec()
+                },
+            )
+            .collect_vec()
+    }
```
Comment (ljeub-pometry) [inline raphtory/src/python/graph/edge.rs:942]: these probably shouldn't collect the values!

File: raphtory/src/python/graph/edge.rs [hunk 4]
```diff
@@ -599,6 +677,174 @@ impl PyEdges {
...
+
+    /// Get the layer names that all edges belong to - assuming they only belong to one layer
+    ///
+    /// Returns:
+    ///   A list of layer names
+    #[getter]
+    fn layer_names(&self) -> Vec<Vec<ArcStr>> {
+        let edges = self.builder.clone();
+        edges().layer_names().map(|e| e.collect_vec()).collect_vec()
+    }
```
Comment (ljeub-pometry) [inline raphtory/src/python/graph/edge.rs:749]: don't collect the result

File: raphtory/src/python/graph/edge.rs [hunk 5]
```diff
@@ -599,6 +677,174 @@ impl PyEdges {
...
+
+    /// Get the start time of all edges
+    ///
+    /// Returns:
+    /// The start time of all edges
+    #[getter]
+    fn start(&self) -> PyGenericIterable {
+        let edges = self.builder.clone();
+        (move || edges().start()).into()
+    }
```
Comment (ljeub-pometry) [inline raphtory/src/python/graph/edge.rs:699]: this should return OptionI64Iterable so you can do things like max and min

File: raphtory/src/python/graph/edge.rs [hunk 6]
```diff
@@ -599,6 +677,174 @@ impl PyEdges {
...
+
+    /// Get the end time of all edges
+    ///
+    /// Returns:
+    /// The end time of all edges
+    #[getter]
+    fn end(&self) -> PyGenericIterable {
+        let edges = self.builder.clone();
+        (move || edges().end()).into()
+    }
```
Comment (ljeub-pometry) [inline raphtory/src/python/graph/edge.rs:719]: same as above, make this return OptionI64Iterable

File: raphtory/src/python/graph/edge.rs [hunk 7]
```diff
@@ -599,6 +677,174 @@ impl PyEdges {
...
+
+    /// Get the end date time of all edges
+    ///
+    /// Returns:
+    ///  The end date time of all edges
+    #[getter]
+    fn end_date_time(&self) -> PyGenericIterable {
+        let edges = self.builder.clone();
+        (move || edges().end_date_time()).into()
+    }
```
Comment (ljeub-pometry) [inline raphtory/src/python/graph/edge.rs:729]: use OptionNaiveDateTimeIterable here so we get the type-specific methods

File: raphtory/src/python/graph/edge.rs [hunk 8]
```diff
@@ -599,6 +677,174 @@ impl PyEdges {
...
+
+    /// Get the start date time of all edges
+    ///
+    /// Returns:
+    /// The start date time of all edges
+    #[getter]
+    fn start_date_time(&self) -> PyGenericIterable {
+        let edges = self.builder.clone();
+        (move || edges().start_date_time()).into()
+    }
```
Comment (ljeub-pometry) [inline raphtory/src/python/graph/edge.rs:709]: this should return OptionNaiveDateTimeIterable

File: raphtory/src/python/graph/edge.rs [hunk 9]
```diff
@@ -599,6 +677,174 @@ impl PyEdges {
         (move || edges().id()).into()
     }
 
+    /// Returns all timestamps of edges, when an edge is added or change to an edge is made.
+    ///
+    /// Returns:
+    ///    A list of timestamps.
+    ///
+
+    fn history(&self) -> PyGenericIterable {
```
Comment (ljeub-pometry) [inline raphtory/src/python/graph/edge.rs:686]: make this return I64VecIterable

File: raphtory/src/python/graph/edge.rs [hunk 10]
```diff
@@ -599,6 +677,174 @@ impl PyEdges {
...
+
+    /// Get the layer name that all edges belong to - assuming they only belong to one layer
+    ///
+    /// Returns:
+    ///  The name of the layer
+    #[getter]
+    fn layer_name(&self) -> PyGenericIterable {
+        let edges = self.builder.clone();
+        (move || edges().layer_name()).into()
+    }
```
Comment (ljeub-pometry) [inline raphtory/src/python/graph/edge.rs:739]: this should return StringIterable

**Review by ljeub-pometry — CHANGES_REQUESTED** [review]
Some return types should use the typed Iterables so we get all the methods. Otherwise looks good to merge.

File: raphtory/src/python/graph/edge.rs (hunk 10 above)
Comment (rachchan) [reply raphtory/src/python/graph/edge.rs:739]: I am returning a custom type OptionArcStringIterable now

**Review by miratepuffin — APPROVED** [review]

**Pull request merged**